src/Builder/Select/QueryExpressionVisitor.php

```diff
@@ -0,0 +1,178 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Solarium\Builder\Select;
+
+use Solarium\Builder\AbstractExpressionVisitor;
+use Solarium\Builder\Comparison;
+use Solarium\Builder\CompositeComparison;
+use Solarium\Builder\ExpressionInterface;
+use Solarium\Builder\Value;
+use Solarium\Exception\RuntimeException;
+
+/**
+ * Query Expression Visitor.
+ *
+ * @author wicliff <wicliff.wolda@gmail.com>
+ */
+class QueryExpressionVisitor extends AbstractExpressionVisitor
+{
+    private const PATTERN = '/( |\+|-|&&|\|\||!|\(|\)|\{|}|\[|]|\^|"|~|\*|\?|:|\/|\\\)/';
+
+    /**
+     * @param \Solarium\Builder\ExpressionInterface $expression
+     *
+     * @throws \Solarium\Exception\RuntimeException
+     *
+     * @return mixed|string
+     */
+    public function walkExpression(ExpressionInterface $expression)
+    {
+        $field = $expression->getField();
+        $value = $expression->getValue()->getValue();
+
+        switch ($expression->getOperator()) {
+            case Comparison::EQ:
+            case Comparison::NEQ:
+                $strValue = $this->valueToString($value, ',', '"');
+
+                if ($value instanceof \DateTime) {
+                    $strValue = sprintf('[%1$s TO %1$s]', $strValue);
+                }
+
+                $not = (Comparison::NEQ === $expression->getOperator()) ? '-' : '';
+
+                return sprintf('%s%s:%s', $not, $field, $strValue);
+            case Comparison::GT:
```

Comment:
In drupal's Search API Solr module we have a very similar logic compared to all these "cases".
We should compare both implementations and isolate it here / move it to Helper in order to leverage it from Search API Solr.

Have a look at https://git.drupalcode.org/project/search_api_solr/blob/8.x-3.x/src/Plugin/search_api/backend/SearchApiSolrBackend.php#L2925

Reply:
I think we might have solved some more edge cases there. In general it would be a good choice to move that stuff to solarium and to leverage it form Search API Solr, regardless if as methods or the new Query Builder.